### PR TITLE
S3655: Boxing

### PR DESCRIPTION
**Pull request by antonioaversa**
Part 5 of task 6 of https://github.com/SonarSource/sonar-dotnet/issues/6794

Previous task: https://github.com/SonarSource/sonar-dotnet/pull/6989

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 1]
```diff
@@ -850,6 +850,70 @@ async Task WithAsyncAwait()
...
+    }
+
+    void NonEmptyNullable()
+    {
+        int? nullable = 42;
+        object implicitBoxed = nullable;
+        _ = (int)implicitBoxed;             // Compliant
+        var explicitBoxed = (object)nullable;
+        _ = (int)explicitBoxed;             // Compliant
+        _ = (int)(object)(int?)(null);      // Compliant
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:880]: Move to `EmptyNullable`

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 2]
```diff
@@ -850,6 +850,70 @@ async Task WithAsyncAwait()
...
+        _ = (int)(object)(null as int?);    // Compliant
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:881]: Delete, already present in `EmptyNullable`

**Review by Tim-Pohlmann — APPROVED** [review]
LGTM, only minor nitpicks.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/EmptyNullableValueAccessTest.cs [hunk 3]
```diff
@@ -61,6 +61,10 @@ public class EmptyNullableValueAccessTest
             .WithOptions(ParseOptionsHelper.FromCSharp8)
             .Verify();
 
+    [TestMethod]
+    public void EmptyNullableValueAccess_Roslyn_NullableContext() =>
+        roslynCS.AddPaths("EmptyNullableValueAccess.NullableContext.cs").WithOptions(ParseOptionsHelper.FromCSharp8).WithConcurrentAnalysis(false).Verify();
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/EmptyNullableValueAccessTest.cs:66]: I think the trick is to add an empty newline at the beginning of the test file, but I'm not sure. Please give it a try.
```suggestion
        roslynCS.AddPaths("EmptyNullableValueAccess.NullableContext.cs").WithOptions(ParseOptionsHelper.FromCSharp8).Verify();
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.NullableContext.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+class NullForgivingOperator
+{
+    void Basics(int? i)
+    {
+        _ = i!.Value;               // Compliant, unknown
+        i = SomeMethod();
+        _ = i!.Value;               // Compliant, unknown
+
+        i = null;
+        _ = i!.Value;               // Noncompliant, empty
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.NullableContext.cs:14]: What is our strategy here? With the "possible null reference" rule we honored the user's wish to overrule us. If this should be the case here too, then this needs to be marked as FP.
What is the "empty" comment referring to? Please use "null" here and below instead.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 5]
```diff
@@ -582,6 +663,256 @@ void AsOperatorWithNonNullLiteral(int? i)
...
+        ReadAndModifyRefParam(iLocal.Value, ref iLocal); // Noncompliant
+
+        nullableField = null;
+        ModifyRefParam(ref nullableField);
+        _ = nullableField.Value;                         // Compliant
+
+        static void ModifyRefParam(ref int? i) => i = null;
+        static void ModifyRefParamAndRead(ref int? i1, int? i2) => i1 = i2;
+        static void ReadAndModifyRefParam(int? i1, ref int? i2) => i2 = i1;
+    }
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:737]: Can we get examples with [`in`-modifier](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/in-parameter-modifier)? One or two simple cases should be enough.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 6]
```diff
@@ -582,6 +663,256 @@ void AsOperatorWithNonNullLiteral(int? i)
...
+
+    void LocalFunctions()
+    {
+        theField = 42;
+        LocalFunctionChangingTheField();
+        _ = theField.Value;          // Compliant, unknown
+
+        theField = null;
+        LocalFunctionChangingTheField();
+        _ = theField.Value;          // Noncompliant, FP, local functions don't reset the state
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:761]: @pavel-mikula-sonarsource: for your attention so we don't miss this case in the hardening phase.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 7]
```diff
@@ -582,6 +663,256 @@ void AsOperatorWithNonNullLiteral(int? i)
...
+        theField = null;
+        MethodChangingTheField();
+        _ = theField.Value;          // Compliant, instance methods reset the state
+
+        theField = 42;
+        MethodNotChangingTheField();
+        _ = theField.Value;          // Compliant, unknown
+
+        void LocalFunctionChangingTheField() => theField = null;
+        void LocalFunctionNotChangingTheField() { }
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:786]:
```suggestion
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 8]
```diff
@@ -582,6 +663,256 @@ void AsOperatorWithNonNullLiteral(int? i)
...
+    void MethodNotChangingTheField() { }
+
+    IEnumerable<int> WithYield()
+    {
+        _ = theField.Value;          // Compliant, unknown
+        yield return 42;
+
+        theField = null;
+        yield return 42;
+        _ = theField.Value;          // Noncompliant, FP, should be unknown
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:800]: @pavel-mikula-sonarsource another one for your attention. Where do we track these? `yield return` should behave exactly like `await` as both turn the method into a state machine with a re-entry at a later point in execution.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.NullableContext.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
+﻿#nullable enable
+
+using System;
+
+class NullForgivingOperator
+{
+    void Basics(int? i)
+    {
+        _ = i!.Value;               // Compliant, unknown
```
[the remark this answers is not in the document]
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.NullableContext.cs:9]: It's not. The PR originally contained a single commit, related to Boxing, and it was targeting `Antonio/S3655-cs9-cs10-step6-4` (the previous part of step 6 of the umbrella issue for S3655 implementation for C#).

Merging `Antonio/S3655-cs9-cs10-syntax-step6-4` into `feature/SE` (done by https://github.com/SonarSource/sonar-dotnet/pull/6996) has made GitHub automatically change the target branch of this PR from `Antonio/S3655-cs9-cs10-syntax-step6-4` to `feature/SE`, which means that now the PR contains all commits from `Antonio/S3655-cs9-cs10-syntax-step6-1` to `Antonio/S3655-cs9-cs10-syntax-step6-4` (which were all merged into `feature/SE` yesterday, since they were all blocked by the review of `Antonio/S3655-cs9-cs10-syntax-step6`, taking longer than all the other reviews).

So in conclusion, you have mostly reviewed commits that have already been reviewed in previous PRs, and which have already been merged into `feature/SE`. Sorry for that :-(

Next time I will try to prevent the situation from happening by putting back in draft state a PR with a dependency on another PR, and also signaling to the reviewer that a rebase is required before the PR can be reviewed using the Changes tab. The problem is that we also use draft to signal not to merge, and we allow early reviews of draft PR, so the only way to avoid such a scenario that I currently see is direct communication, issuing a warning.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 10]
```diff
@@ -850,6 +850,69 @@ async Task WithAsyncAwait()
...
+        _ = (int)(object)(int?)(null);      // Compliant
+        _ = (int)(object)(null as int?);    // Compliant
+    }
+
+    void NullableRoundTrip()
+    {
+        int? nullable = null;
+        object implicitBoxed = nullable;
+        int? unboxedNullable = (int?)implicitBoxed;
+        _ = unboxedNullable.Value;          // Noncompliant, empty
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:871]: Please add a non-empty counter-example where `int? nullable = ...;` is initialized with a value.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.NullableContext.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.NullableContext.cs]: As discussed offline with @martin-strecker-sonarsource, I have rebased onto the new `feature/SE`, and I am going to address the comments on code already merged in a dedicated PR.
This PR is back to the state it was before the auto-change of base branch by GitHub.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 11]
```diff
@@ -850,6 +850,69 @@ async Task WithAsyncAwait()
...
+    {
+        foreach (object boxed in new int?[] { null })
+        {
+            _ = (int)boxed;                 // Compliant
+        }
+    }
+
+    void CollectionImplicitBoxing()
+    {
+        foreach (var boxed in new object[] { null as int? })
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:900]: We already have this example at another place right:
```cs
        foreach (int boxed in new int?[] { null })
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Looks good. One roundtrip example should be added.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/EmptyNullableValueAccessTest.cs (hunk 3 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/EmptyNullableValueAccessTest.cs]: It does work without `WithConcurrentAnalysis(false)`, even not adding the empty newline at the beginning of the test file.

So, I have removed `WithConcurrentAnalysis(false)` in https://github.com/SonarSource/sonar-dotnet/pull/7006 (new PR for all issues detected on this PR but on code already merged into `feature/SE`).

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.NullableContext.cs (hunk 4 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.NullableContext.cs]: My understanding, deriving from a discussion we had during one of our sync meeting, is the following.

In the rule checking about possible null references, we trigger an issue every time we find a dereference of something which can be `null`. If we don't know for certain (i.e. unknown value), we trigger, because the possibility of dereference is there.

In this rule, however, we only trigger when we know for certain that there is a path where the variable is `null`. In the case above, for example, `i` is assigned to be `null` right before value accessing, so we know for certain that an `InvalidOperationException` will be produced as a result of calling `Value` on `i`, no matter if the null forgiving operator is used or not.

So we basically ignore and ask the user to protect against the scenario where the variable surely get a `null`.

An objection to this line of reasoning may be that a simple check of `HasValue` would trigger the rule:
```cs
void Method(int? i)
{
    if (i.HasValue) { }
    _ = i.Value; // Noncompliant, there's a path where i is empty
}
```
whereas
```
void Method(int? i)
{
    _ = i.Value; // Compliant, unknown
}
```

My understanding (and I would ask @pavel-mikula-sonarsource to back up or dismiss the reasoning) is that if the user has made a previous call to `i.HasValue`, he has somewhat acknowledged the fact that `i` can be empty. So we then ask him to be consistent with that observation, and protect his access to `i.Value` with a check `i.HasValue` or `i != null`.

Concerning the use of `, empty` after `Noncompliant`: it refers to the nullable, which is empty. I have used the word empty, rather than `null`, because the rule is called `EmptyNullableValueAccess` and also because I find misleading saying that a `struct` can be `null`. It's rather the opposite: `null` is converted by the compiler into a `Nullable<T>` struct with `hasValue == false`.

Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.NullableContext.cs]: We should honor the null forgiving operator. In some cases, our engine is wrong, and the user should be able to make the rule shut up.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 7 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: Removed these two unused local functions in https://github.com/SonarSource/sonar-dotnet/pull/7006 (new PR for all issues detected on this PR but on code already merged into `feature/SE`).

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 10 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:871]: Added. I have also added the case where `nullable` is a method parameter, and as such, its value is unknown.

This way we should have all basic scenarios: empty, non-empty, and unknown.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 6 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: We can ignore this problem for now, because it was not supported in the old engine too.

The evil plan is to actually execute the body of local functions in [MMF-2228](https://jira.sonarsource.com/browse/MMF-2228)

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 8 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: We don't (yet). 
I think there's a difference between await and yield return. Await can mess up your local variables and parameters, while yield return can mess up only external state (fields), like a local method invocation.

So I'd propose to threat it like a local method invocation to reset only some state.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.NullableContext.cs (hunk 4 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.NullableContext.cs]: To correct @antonioaversa understanding: Both S3655 and S2259 behave the same in principle. If user indicated possible null (for example by asking `nullable.HasValue` or `arg is null` on the other rule), we assume possible null and will raise. S2259 does not raise if we don't know to reduce noise.

As `!` means "trust me, it works (and I won't blame you)", I'm fine with keeping the behavior symmetrical and support bang here too.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 11 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:919]: I think we should have test cases for for each for the following cases:

### explicit conversion from nullable to non-nullable

* explicit conversion from int? to int (e.g. `foreach(int i in new int?[] { 1, null })`)
* lifted explicit conversions (e.g. `foreach(int i in new long?[] { 1, null })`). Here there are actually two explicit conversions at stake: one form nullable to non-nullable and one from wider type (long) to narrow type (int).

These are the cases I asked for originally.

### boxing conversions from nullable to object

`foreach (object o in new int?[] { null })`

These are [implicit conversions](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/conversions#1028-implicit-reference-conversions). This kind of operation should always succeed, so I don't see the need for much test coverage here.

### unboxing into nullable

`foreach (int? i in new object[] { null, 1, "Boom" })`

This has nothing to do with the rule itself. No need to cover it with tests.

### stackalloc

I don't see how this makes a difference. It is just another kind of location to enumerate. You can also use other kinds of pattern-based foreach collections, like ImmutableArrays Enumerator. It doesn't make a difference for the test cases.

In general, I would look at the C# spec rather than the IL generated. The IL generated needs to follow the spec behavior-wise. I don't see any violations from that in the examples you have given.

The sections needed are
* [Conversion from a nullable value type to an underlying type](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types#conversion-from-a-nullable-value-type-to-an-underlying-type)
* [Lifted operators](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types#lifted-operators)
* [foreach](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/iteration-statements#the-foreach-statement)

Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:919]: Ok, good, it's clear now, thanks for this analysis! 

So:

### explicit conversion from nullable to non-nullable
Added a new method `ForeachExplicitConversion` with the two examples

### boxing conversions from nullable to object
The conversion indeed succeeds, but the cast in the body of the foreach does not.
`(int)boxed` fails with `NullReferenceException` at the first item of the collection which is `null`.
I find non-obvious at first glance that:
- if boxing happens, you end up with a `NullReferenceException`, raised by the runtime in the boxing operation, and S3655 does not kick in;
- whereas, if boxing doesn't happen, you end up with an `InvalidOperationException`, raised by the framework in `Nullable<T>.Value`, and S3655 does kick in.
I wanted to document this difference in behavior, by keeping these two test scenarios `ForeachImplictConversion` and `ForeachExplicitConversion`.

### unboxing into nullable
Agreed, I haven't added it.

### stackalloc
Indeed, it just change the memory layout. 
It just seemed interesting, since stack allocation in general can affect boxing. 
Removed.

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.NullableContext.cs (hunk 4 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.NullableContext.cs]: Great, thanks for the explanation. Support of this will come in a dedicated PR: https://github.com/SonarSource/sonar-dotnet/pull/7015
